core/gather/driver.js

```diff
@@ -67,9 +68,13 @@ class Driver {
     return this._page.url();
   }
 
+  isConnected() {
+    return this._connected;
```

Comment:
Seems redundant with `throwingSession`? e.g. `isConnected` could be `return .defaultSession !== throwingSession` as long as `defaultSession` was set back to `throwingSession` in `disconnect`

Reply:
It's necessary to mark the driver as "not connected" *before* disposing of the session. We would need to temporarily hold on to the `defaultSession` after setting it back to `throwingSession` in `disconnect` so we can properly dispose it. This way also makes more semantic sense.

Reply:
Is it necessary or is it better to do it before? It doesn't appear to be reentrant, but I could see the argument that one day it could be and better to be prepared.

> We would need to temporarily hold on to the `defaultSession` after setting it back to `throwingSession` in `disconnect` so we can properly dispose it

to clarify, do you mean like
```js
if (defaultSession === throwingSession) return;
const oldSession = this.defaultSession;
this.defaultSession = throwingSession;

await this._targetManager?.disable();
await oldSession.dispose();
```
?

> This way also makes more semantic sense.

I agree, but we have this preexisting big chunk of code for `throwingSession` that duplicates 90% of the functionality of this new variable and could be 100% with a code change. We could get rid of `throwingSession` instead.

A long time ago we discussed making the FR driver pseudo-RAII like the legacy driver and there wouldn't be the pre/post connection state to deal with, but that ship has probably long sailed.

Reply:
> to clarify, do you mean like

Ya. I guess they're both imperfect, your suggestion makes sense though.